### ISSUE TITLE
Improve redis memory/disk usage for sdk-scheme-adapter

### ISSUE THREAD

**Issue by lewisdaly** [mojaloop/project#1134]
**Summary**:
When running load tests on a Mojaloop system using `mojaloop-simulator`, the required redis container runs out of disk space.

**Tasks**
- [ ] Verify that the redis cache in mojaloop-simulator is causing the underlying issue with disk usage
- [ ] establish a strategy to expire objects from the cache, either deleting them manually, or setting a configurable expiry time:  https://redis.io/commands/expire
- [ ] add integration tests to mojaloop-simulator, and verify fixes

**Severity**:
Medium

**Priority**:
Medium

**Expected Behavior**
The `mojaloop-simulator` should run indefinitely without running out of disk space.

**Steps to Reproduce**
  1. In the perf environment, run many JMeter tests
  2. check disk usage of mojaloop-simulator instances running with docker-compose over time, and observe that disks are filling up
  3.

**Specifications**
  - Component (if known): mojaloop-simulator
  - Version: v8.6.4
  - Platform: AWS
  - Subsystem: sdk-scheme-adapter
  - Type of testing: Performance testing
  - Bug found/raised by: 

**Notes**:
We currently think this is because the `sdk-scheme-adapter`, which uses redis as a cache, only ever calls `SET` and not `DELETE`. For every request to the scheme adapter, a new entry is added to the cache, doesn't get deleted, which presumably is impacting both memory and disk perfomance.

- Severity when opened: Medium
- Priority when opened: Medium

**Comment by lewisdaly** [comment]
I think I've managed to confirm this issue when running the `mojaloop-simulator` locally:


1. This log from redis instances:
```
redis-sim_1              | 1:M 19 Dec 2019 11:56:23.119 * 100 changes in 300 seconds. Saving...
redis-dfsp_1             | 1:M 19 Dec 2019 11:56:23.115 * 100 changes in 300 seconds. Saving...
redis-sim_1              | 1:M 19 Dec 2019 11:56:23.127 * Background saving started by pid 13
redis-dfsp_1             | 1:M 19 Dec 2019 11:56:23.128 * Background saving started by pid 16
redis-sim_1              | 13:C 19 Dec 2019 11:56:23.166 * DB saved on disk
redis-dfsp_1             | 16:C 19 Dec 2019 11:56:23.166 * DB saved on disk
redis-sim_1              | 13:C 19 Dec 2019 11:56:23.167 * RDB: 0 MB of memory used by copy-on-write
redis-dfsp_1             | 16:C 19 Dec 2019 11:56:23.168 * RDB: 0 MB of memory used by copy-on-write
redis-sim_1              | 1:M 19 Dec 2019 11:56:23.236 * Background saving terminated with success
redis-dfsp_1             | 1:M 19 Dec 2019 11:56:23.235 * Background saving terminated with success
```

implies that if there are many changes to the cache, then it will be written to the `/data/dump.rdb` file.

2. Check this file size with:
```bash
$ docker exec -it src_redis-sim_1 sh -c "ls -lah"
total 1108
drwxr-xr-x    2 redis    redis       4.0K Dec 19 11:56 .
drwxr-xr-x    1 root     root        4.0K Dec 19 11:00 ..
-rw-r--r--    1 redis    redis       1.1M Dec 19 11:56 dump.rdb
```


3. After another 300 requests:
```
-rw-r--r--    1 redis    redis       1.2M Dec 19 12:17 dump.rdb
```

Summary:
- every 600 seconds, a new dump file is saved
- After 300 new requests, this dump file grew 0.1 MB
  - so with some really rough estimating, we can estimate 1 request = 0.000333333 MB
  - and 1,000,000 Requests ~= 300 MB

**Comment by lewisdaly** [comment]
As per @bushjames' comments on: https://github.com/mojaloop/sdk-scheme-adapter/pull/122 A better option would be to configure the test redis instances as caches, since this wouldn't add a new setting that might compromise a production instance of the `sdk-scheme-adapter`

**Comment by lewisdaly** [comment]
I've got another approach working by setting the`maxmemory` and `maxmemory-policy` settings in `redis.conf`.

Here's a summary of my findings:


0. Check file sizes:
```bash
docker exec -it redis-sim sh -c "ls -lah"
docker exec -it redis-dfsp sh -c "ls -lah"

```

redis-sim:dump.rdb  -> no file
redis-dfsp:dump.rdb -> no file


1. Run the load test:
```bash
artillery run redis_test.yml
```

Notice any logs:

```
redis-sim                | 1:M 13 Jan 2020 03:13:58.079 * 10 changes in 300 seconds. Saving...
redis-sim                | 1:M 13 Jan 2020 03:13:58.110 * Background saving started by pid 18
redis-dfsp               | 1:M 13 Jan 2020 03:13:58.066 * 10 changes in 300 seconds. Saving...
redis-dfsp               | 1:M 13 Jan 2020 03:13:58.109 * Background saving started by pid 12
redis-dfsp               | 12:C 13 Jan 2020 03:13:58.145 * DB saved on disk
redis-dfsp               | 12:C 13 Jan 2020 03:13:58.146 * RDB: 0 MB of memory used by copy-on-write
redis-sim                | 18:C 13 Jan 2020 03:13:58.141 * DB saved on disk
redis-sim                | 18:C 13 Jan 2020 03:13:58.142 * RDB: 0 MB of memory used by copy-on-write
redis-sim                | 1:M 13 Jan 2020 03:13:58.230 * Background saving terminated with success
redis-dfsp               | 1:M 13 Jan 2020 03:13:58.261 * Background saving terminated with success
```


2. Check the dump size:

```bash
docker exec -it redis-sim sh -c "ls -lah"
docker exec -it redis-dfsp sh -c "ls -lah"

```

redis-sim:dump.rdb  -> 1.2K
redis-dfsp:dump.rdb ->  168.1K

3. Run again:
```bash
artillery run redis_test.yml
```

4. Check again:

```bash
docker exec -it redis-sim sh -c "ls -lah"
docker exec -it redis-dfsp sh -c "ls -lah"

```

redis-sim:dump.rdb  -> 1.2K
redis-dfsp:dump.rdb ->  168.1K

Summary:
- it seems that the initial dump gets created, but not updated (this is likely because our test is the same request over and over again)
- that said, the memory limit looks like it is expring object in the cache.
    - We know this because the previous set of tests had a dump size of 1.2MB after 1 load test

**Comment by lewisdaly** [comment]
@mdebarros I think the approach from here should be to add our own custom `redis.conf` file in the helm charts for the mojaloop-simulator.

**Comment by mdebarros** [comment]
> @mdebarros I think the approach from here should be to add our own custom `redis.conf` file in the helm charts for the mojaloop-simulator.

Didn't this PR resolve this issue since the cache expires based on the config --> https://github.com/mojaloop/sdk-scheme-adapter/pull/122/files#diff-c96817cdaa724a259240571a67bca907R48
- CACHE_SHOULD_EXPIRE
- CACHE_EXPIRY_SECONDS
?

**Comment by lewisdaly** [comment]
No. @bushjames was worried that by introducing these env variables, we would encourage them to be used in production, which they absolutely should not be. 

I think changing the config on redis is a more appropriate solution anyway. I've got a `redis.conf` file ready to go, I just need to know where to put it in helm (or wherever the `mojaloop-simulator` helm charts live)

**Comment by elnyry** [comment]
@lewisdaly there were some changes made in in the sdk-scheme-adapter recently and this may have been covered there.. Can you please confirm, so that we can unblock this?

**Comment by lewisdaly** [comment]
I don't think the changes in the `sdk-scheme-adapter` will unblock it. But the changes in Helm to use the mojaloop-simulator _will_.

As @bushjames noted in this PR: https://github.com/mojaloop/sdk-scheme-adapter/pull/122#pullrequestreview-341061522

We should add these settings in a `redis.conf` file since they should _never_ be used in a production deployment.

**Comment by lewisdaly** [comment]
Seems out of scope now. Please reopen if I'm wrong.
